**Problem 2**

In [1]:
# import the necessary packages
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.outliers_influence import variance_inflation_factor

%matplotlib inline

In [2]:
# input the Honda Accord dataset
accord = pd.read_csv('Accord-242A-Spring24.csv')

**a)**

In [3]:
#split the training and testing dataset; cutoff point is 2018
accord_train = accord[accord['Year'] <= 2018]
accord_test = accord[accord['Year'] > 2018]

len(accord_train), len(accord_test)

(60, 59)

In [4]:
# Choose the features to be used
cols_1 = ['Unemployment', 'AccordQueries', 'CPIAll', 'CPIEnergy', 'MilesTraveled']
X_train_1 = accord_train[cols_1]
y_train = accord_train['AccordSales']

# Add an intercept
X_train_1 = sm.add_constant(X_train_1)

# Fit the data to the model
model_1 = sm.OLS(y_train, X_train_1).fit() #ordinary least square
print(model_1.summary())

# Calculate Variance Inflation Factor for each explanatory variable
def VIF(df, columns):
    values = sm.add_constant(df[columns]).values
    num_columns = len(columns)+1
    vif = [variance_inflation_factor(values, i) for i in range(num_columns)]
    return pd.Series(vif[1:], index=columns)

VIF(X_train_1, cols_1)

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.254
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     3.683
Date:                Mon, 12 Feb 2024   Prob (F-statistic):            0.00612
Time:                        15:27:30   Log-Likelihood:                -595.60
No. Observations:                  60   AIC:                             1203.
Df Residuals:                      54   BIC:                             1216.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          1.735e+05   1.49e+05      1.164

Unemployment     31.683094
AccordQueries     1.939028
CPIAll           45.304922
CPIEnergy        12.340357
MilesTraveled    18.510210
dtype: float64

In [5]:
# Choose the features to be used
cols_2 = ['AccordQueries', 'CPIAll', 'CPIEnergy', 'MilesTraveled']
X_train_2 = accord_train[cols_2]
y_train = accord_train['AccordSales']

# Add an intercept
X_train_2 = sm.add_constant(X_train_2)

# Fit the data to the model
model_2 = sm.OLS(y_train, X_train_2).fit() #ordinary least square
print(model_2.summary())

# calculate Variance Inflation Factor for each explanatory variable
VIF(X_train_2, cols_2)

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.252
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     4.636
Date:                Mon, 12 Feb 2024   Prob (F-statistic):            0.00269
Time:                        15:27:30   Log-Likelihood:                -595.69
No. Observations:                  60   AIC:                             1201.
Df Residuals:                      55   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          1.171e+05    3.8e+04      3.084

AccordQueries     1.918686
CPIAll           14.065391
CPIEnergy         4.634036
MilesTraveled    17.765021
dtype: float64

In [6]:
# Choose the features to be used
cols_3 = ['AccordQueries', 'CPIAll', 'CPIEnergy']
X_train_3 = accord_train[cols_3]
y_train = accord_train['AccordSales']

# Add an intercept
X_train_3 = sm.add_constant(X_train_3)

# Fit the data to the model
model_3 = sm.OLS(y_train, X_train_3).fit() #ordinary least square
print(model_3.summary())

# calculate Variance Inflation Factor for each explanatory variable
VIF(X_train_3, cols_3)

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.227
Model:                            OLS   Adj. R-squared:                  0.185
Method:                 Least Squares   F-statistic:                     5.475
Date:                Mon, 12 Feb 2024   Prob (F-statistic):            0.00227
Time:                        15:27:30   Log-Likelihood:                -596.69
No. Observations:                  60   AIC:                             1201.
Df Residuals:                      56   BIC:                             1210.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          1.419e+05   3.36e+04      4.224

AccordQueries    1.899305
CPIAll           1.816792
CPIEnergy        1.075859
dtype: float64

**(i)**

The linear regression equation (approximated to two decimal places) produced by the final selected model is monthly AccordSales = 141,900.00 + 245.21* AccordQueries - 610.92* CPIAll + 67.03* CPIEnergy in the US. The monthly Accord sales is 141,900 in the US if all the independent varibales are zero. One more Accord query on Google holding all other features constant will lead to approximately 245 more sales of Accord in a given month. A unit increase in the consumer price index for all products holding all other features constant will lead to approximately 610 fewer sales of Accord in a given month. A unit increase in the consumer price index for the energy sector holding all other features constant will lead to approximately 67 more sales of Accord in a given month. 

**ii)**

model_1 includes all five features, Unemployment, AccordQueries， CPIAll, CPIEnergy, and MilesTraveled. Except AccordQueries, all other four features have VIF higher than 5. Given Unemployment has the highest p-value around 0.7, I decided to remove Unemployment in model_2 and see if the model performance improves.

model_2 includes AccordQueries, CPIAll, CPIEnergy, and MilesTraveled. CPIAll and MilesTraveled have VIF higher than 5. Given MilesTraveled is the only feature that has the p-value higer than 0.05, I decided to remove MilesTraveled in model_3 and see if the model performance improves. Comparing to model_1, model_2 has higher adjusted R-squared at 0.198 vs 0.185.

model_3 includes AccordQueries, CPIAll, and CPIEnergy. All the VIFs are lower than 5. Although CPIEnergy has the p-value slightly higer than 0.05, I decided to retain it. Comparing to model_2, although model_3 has lower adjusted R-squared at 0.185 vs 0.198, I decided to choose model_3 as the final selected model given reasonable values of VIF, p-value, and adjusted R-suqared comparatively.

**iii)**

The number of Accord Queries and the consumer price index for the energy sector in the US have a positive impact on the sales when these two independent variables increase. It makes sense as a higher number of Accord Queries in the US indicates there is a higher interest particularly in Accord. However, the reasoning is unclear that the consumer price index for the energy sector has a positive impact. From commen sense, if the gasoline price goes up, which is included in the consumer price index for the energy sector, consumers should demand less gas-powered vehicles but are more likely to switch to electric vehicles. Accord has both gas-powered version and hybrid-electric version and the decomposition of these two versions are not available from the dataset.

On the other hand, the consumer price index for all products has a negative impact on the sales when the independent varibales increases. It makes sense as a higher consumer price index for all products (i.e., inflation) usually indicates an economic recession. Consequently, consumers have less real disposable income and they are less likely to buy Accord and any other goods and services.

**iv)**

The model performance is very poor given the R-squared is only 0.227 and the adjusted R-squared is only 0.185.

**b)**

In [7]:
# Fit the data to the model
model_4 = smf.ols(formula='AccordSales ~ MonthFactor + AccordQueries + Unemployment + CPIAll + CPIEnergy + MilesTraveled', 
                  data=accord_train).fit() #ordinary least square
print(model_4.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.654
Method:                 Least Squares   F-statistic:                     7.982
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           2.66e-08
Time:                        15:27:30   Log-Likelihood:                -563.04
No. Observations:                  60   AIC:                             1160.
Df Residuals:                      43   BIC:                             1196.
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

**i)**

The linear regression equation (approximated to two decimal places) produced by the final selected model is monthly AccordSales = 97,750 + 8697.67* [if it is August] + 3256.30* [if it is December] - 4825.49* [if it is February] - 8189.76* [if it is January] + 3894.86* [if it is July] +1536.54* [if it is June] + 430.15* [if it is March] + 5573.22* [if it is May] - 1493.65* [if it is November] - 25.63* [if it is October] + 3046.17* [if it is September] + 762.82* Unemployment + 10.66* AccordQueries - 639.81* CPIAll + 67.52* CPIEnergy + 0.25* MilesTraveled. 

The monthly AccordSales in April is 97,750 if all other features are zero.

The coefficient of each of the MonthFactor dummy varibale is the monthly AccordSales difference comparing to the month of April holding all other features constant. For example, if it is August, the August dummy varibale is 1 and all other MonthFactor dummy varibales are all 0. The monthly AccordSales in August will be around 8697 higher than that in April, leading to a monthly sale of 106,447 in August. 

A percentage increase in the unemployment rate holding all other features constant will lead to approximately 762 more sales of Accord in a given month. One more Accord query on Google holding all other features constant will lead to approximately 10 more sales of Accord in a given month. A unit increase in the consumer price index for all products holding all other features constant will lead to approximately 639 fewer sales of Accord in a given month. A unit increase in the consumer price index for the energy sector holding all other features constant will lead to approximately 67 more sales of Accord in a given month. A unit increase in the MilesTraveled holding all other features constant will lead to approximately 0.2 more sales of Accord in a given month.

**ii)**

The training set R-squared for this model is 0.748. Significant varibales (i,e, p-value is less than 0.05) include MonthFactor[T.August], MonthFactor[T.January], and MonthFactor[T.May].

**iii)**

I think adding the independent varibale MonthFactor improves the quality of the model as the R-squared of model_4 improves signifantly from model_3, 0.748 vs. 0.227.

**iv)**

We can model seasonality according to quarters instead of months, i.e., grouping months into quarter 1, 2, 3, and 4.

**c)**

In [8]:
# Fit the data to the model
model_5 = smf.ols(formula='AccordSales ~ MonthFactor + Unemployment + CPIAll + CPIEnergy + MilesTraveled', 
                  data=accord_train).fit() #ordinary least square
print(model_5.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.748
Model:                            OLS   Adj. R-squared:                  0.662
Method:                 Least Squares   F-statistic:                     8.711
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           8.74e-09
Time:                        15:27:30   Log-Likelihood:                -563.05
No. Observations:                  60   AIC:                             1158.
Df Residuals:                      44   BIC:                             1192.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [9]:
# Fit the data to the model
model_6 = smf.ols(formula='AccordSales ~ MonthFactor + Unemployment + CPIAll + CPIEnergy', 
                  data=accord_train).fit() #ordinary least square
print(model_6.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.745
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     9.388
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           3.53e-09
Time:                        15:27:30   Log-Likelihood:                -563.42
No. Observations:                  60   AIC:                             1157.
Df Residuals:                      45   BIC:                             1188.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [10]:
# Fit the data to the model
model_7 = smf.ols(formula='AccordSales ~ MonthFactor + Unemployment + CPIAll', 
                  data=accord_train).fit() #ordinary least square
print(model_7.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.744
Model:                            OLS   Adj. R-squared:                  0.672
Method:                 Least Squares   F-statistic:                     10.31
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           1.10e-09
Time:                        15:27:30   Log-Likelihood:                -563.47
No. Observations:                  60   AIC:                             1155.
Df Residuals:                      46   BIC:                             1184.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [11]:
# Fit the data to the model
model_8 = smf.ols(formula='AccordSales ~ MonthFactor + Unemployment', 
                  data=accord_train).fit() #ordinary least square
print(model_8.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.736
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     10.94
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           6.20e-10
Time:                        15:27:30   Log-Likelihood:                -564.40
No. Observations:                  60   AIC:                             1155.
Df Residuals:                      47   BIC:                             1182.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

AccordQueries was removed from model_4 to form model_5 as it has the second highest p-value at 0.942 (MonthFactor[T.October] has highest p-value at 0.942) comparing with other features. model_5 has R-squared at 0.748 and adjusted R-squared at 0.662.


MilesTraveled was removed from model_5 to form model_6 as it has a high p-value at 0.463 comparing with other features and the coefficient is close to zero. model_6 has R-squared at 0.745 and adjusted R-squared at 0.666.


CPIEnergy was removed from model_6 to form model_7 as it has the highest p-value other than some of the MonthFactor dummy variables comparing with other features. model_7 has R-squared at 0.744 and adjusted R-squared at 0.672.


CPIAll was removed from model_7 to form model_8 as it has the highest p-value other than some of the MonthFactor dummy variables comparing with other features. model_8 has R-squared at 0.736 and adjusted R-squared at 0.669.


As a result, model_6, which includes MonthFactor, Unemployment, CPIAll, and CPIEnergy as features, is the final model selected given it has reasonable R-squared and adjusted R-squared among the four models compared above. Although it is not the simpliest model, it retains good balance between simplicity and model fittness.

In [12]:
def OSR2(model, df_train, df_test, dependent_var):

    y_test = df_test[dependent_var]
    y_pred = model.predict(df_test)
    SSE = np.sum((y_test - y_pred)**2)
    SST = np.sum((y_test - np.mean(df_train[dependent_var]))**2)

    return 1 - SSE/SST

In [13]:
OSR2(model_6, accord_train, accord_test, 'AccordSales')

0.5658545696406424

The training set R-squared is 0.745 and the OSR-squared is 0.566. The OSR-squared is lower than the R-squared, which makes sense as the OLS regression model is fitted to the training data and usually out-of-sample prediction has a lower R-squared. However, the OSR-squared value of 0.566 is at a concerning level as it indicates the regression model has limited predictibility. The significant drop is probably due to the pandemic, which had a longlasting period from 2020 to 2023. The inflation and unemployement rate were unusually high ihe US, affecting Unemployment, CPIAll, and CPIEnergy. The training set does not include any similar scenario in the dataset, possibility leading to the poor prediction of a rare event.

**d)**

In [14]:
DFF = pd.read_csv('DFF.csv')
DFF.head()

DATE       DFF
0  2014-01-01  0.071613
1  2014-02-01  0.066429
2  2014-03-01  0.078065
3  2014-04-01  0.090333
4  2014-05-01  0.087097

In [15]:
accord_new = accord 
accord_new['DFF']= DFF['DFF']
accord_new.head()

MonthNumeric MonthFactor  Year  AccordSales  Unemployment  AccordQueries  \
0             1     January  2014        20604           6.6             69   
1             2    February  2014        24622           6.7             74   
2             3       March  2014        33962           6.7             79   
3             4       April  2014        34124           6.2             74   
4             5         May  2014        39637           6.3             75   

    CPIAll  CPIEnergy  MilesTraveled       DFF  
0  235.288    250.340         246531  0.071613  
1  235.547    249.925         249499  0.066429  
2  236.028    249.961         251120  0.078065  
3  236.468    249.864         251959  0.090333  
4  236.918    249.213         252289  0.087097

In [16]:
#split the training and testing dataset; cutoff point is 2018
accord_train_new = accord_new[accord_new['Year'] <= 2018]
accord_test_new = accord_new[accord_new['Year'] > 2018]

In [17]:
# Fit the data to the model
model_9 = smf.ols(formula='AccordSales ~ MonthFactor + Unemployment + CPIAll + CPIEnergy + DFF', 
                  data=accord_train_new).fit() #ordinary least square
print(model_9.summary())

                            OLS Regression Results                            
Dep. Variable:            AccordSales   R-squared:                       0.747
Model:                            OLS   Adj. R-squared:                  0.661
Method:                 Least Squares   F-statistic:                     8.677
Date:                Mon, 12 Feb 2024   Prob (F-statistic):           9.26e-09
Time:                        15:27:30   Log-Likelihood:                -563.13
No. Observations:                  60   AIC:                             1158.
Df Residuals:                      44   BIC:                             1192.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

In [18]:
OSR2(model_9, accord_train_new, accord_test_new, 'AccordSales')

0.08004426754171201

Monthly Average Fed Fund Effective Rate(%) (i.e., DFF) from January 2014 to November 2023 is added as an additional feature to the final model from part (c). I hypothesize it might be related to Honda sales as the Fed Fund Rate level impacts personal loan rate in general, which would likely to affect people's incentive to buy cars. In particular, a higher Fed Fund Rate will likely to increase personal loan rate, which makes it more expensive to borrow from the banks and more difficult to find a cheaper refinancing option.

The new model, model_9, including DFF in addtion to the features from the final model, model_6, from part (c) has a slightly better R-squared value at 0.747 vs. 0.745. However, the OSR-squared shows a significnt drop from 0.566 to 0.080, indicating that DFF is not providing any new predictive value but worsening the out-of-sample model performance. The p-value of DFF is 0.520, which is much higher than 0.05, highlighting that it is not a statistically significant feature. The coefficient of -2097.81 is align with my expectation of the feature.

The possible reasons to explain the phenomenum are: 1) model_9 has higher multicollinearity than model_6, i.e., DFF is highly correlated with unemployment and/or CPI; 2) the DFF in the testing data is much unusual as in the training data given the FED consecutively raised the DFF to control the inflation of the US during the pandemic, so the black swan event limited the out-of-sample predictibility.